File: cpp/Divide_an_Array_Into_Subarrays_With_Minimum_Cost_II_3013.cpp
class Solution {
public:

    long long minimumCost(vector<int>& nums, int k, int dist) {
        long long glbSum = 0LL, curSum = 0LL;
        multiset <int> kSmallest;
        multiset <int> rest;
        for (int i = 1; i < k; i++) {
            curSum += nums[i];
            kSmallest.insert(nums[i]);
        }
        glbSum = curSum;
        int l = 1, r = k;
        while (r < nums.size()) {
            auto itMin = prev(kSmallest.end());
            if (nums[r] < *itMin) {
                rest.insert(*itMin);
                curSum -= *itMin;
                kSmallest.erase(itMin);
                kSmallest.insert(nums[r]);
                curSum += nums[r];
            }
            else {
                rest.insert(nums[r]);
            }
            if (r - l > dist) {
                if (kSmallest.find(nums[l]) != kSmallest.end()) {
                    auto itNum = kSmallest.find(nums[l]);
                    curSum -= nums[l];
                    kSmallest.erase(itNum);
                    auto itSmallest = rest.begin();
                    curSum += *itSmallest;
                    kSmallest.insert(*itSmallest);
                    rest.erase(itSmallest);
                }
                else {
                    rest.erase(rest.find(nums[l]));
                }
                l++;
            }
            r++;
            glbSum = min(glbSum, curSum);
        }
        return nums[0] + glbSum;
    }
};

File: python/Divide_an_Array_Into_Subarrays_With_Minimum_Cost_II_3013.py
class Solution:
    def minimumCost(self, nums: List[int], k: int, dist: int) -> int:
        kSmallest, rest = SortedList(nums[1:k]), SortedList()
        curSum, glbSum = sum(kSmallest), sum(kSmallest)
        l, r = 1, k

        while r < len(nums):
            if nums[r] < kSmallest[-1]:
                maxSmallest = kSmallest.pop()
                curSum -= maxSmallest
                rest.add(maxSmallest)
                kSmallest.add(nums[r])
                curSum += nums[r]
            else:
                rest.add(nums[r])
            if r - l > dist:
                if nums[l] <= kSmallest[-1] and nums[l] in kSmallest:
                    kSmallest.remove(nums[l])
                    curSum -= nums[l]
                    newVal = rest.pop(0)
                    kSmallest.add(newVal)
                    curSum += newVal
                else:
                    rest.remove(nums[l])
                l += 1
            r += 1
            glbSum = min(glbSum, curSum)
        
        return nums[0] + glbSum
